services/repository/contributors_graph.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package repository
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	repo_model "code.gitea.io/gitea/models/repo"
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+	api "code.gitea.io/gitea/modules/structs"
+	util "code.gitea.io/gitea/modules/util"
+)
+
+// CreateWeeks converts list of sundays to list of *api.WeekData
+func CreateWeeks(sundays []int64) []*api.WeekData {
+	var weeks []*api.WeekData
+	for _, week := range sundays {
+		weeks = append(weeks, &api.WeekData{
+			Week:      week,
+			Additions: 0,
+			Deletions: 0,
+			Commits:   0,
+		},
+		)
+	}
+	return weeks
+}
+
+// GetContributorStats returns contributors stats for git commits for given revision or default branch
+func GetContributorStats(ctx context.Context, repo *repo_model.Repository, revision string) (map[string]*api.ContributorData, error) {
+	gitRepo, closer, err := git.RepositoryFromContextOrOpen(ctx, repo.RepoPath())
+	if err != nil {
+		return nil, fmt.Errorf("OpenRepository: %w", err)
+	}
+	defer closer.Close()
+
+	if len(revision) == 0 {
+		revision = repo.DefaultBranch
+	}
+	extendedCommitStats, err := gitRepo.ExtendedCommitStats(revision)
+	if err != nil {
+		return nil, fmt.Errorf("ExtendedCommitStats: %w", err)
+	}
+
+	layout := "2006-01-02"
```

Comment:
hmm isn't there some const in the time package already for it?

Reply:
[fixed](https://github.com/go-gitea/gitea/pull/27882/commits/6ebc6ce27f3679991d11b4d2a6c606aca5ed7441)